lib/bgm.dart

```diff
@@ -0,0 +1,101 @@
+import 'dart:io';
+
+import 'package:audioplayers/audioplayers.dart';
+import 'package:flame/flame.dart';
+import 'package:flutter/widgets.dart';
+
+/// The looping **b**ack**g**round **m**usic class.
+///
+/// This class helps with looping background music management that reacts to
+/// application lifecycle state changes. On construction, the instance is added
+/// as an observer to the [WidgetsBinding] instance. A [dispose] function is
+/// provided in case this functionality needs to be unloaded but the app needs
+/// to keep running.
+class Bgm extends WidgetsBindingObserver {
+  AudioPlayer audioPlayer;
+  bool isPlaying = false;
+
+  Bgm() {
+    WidgetsBinding.instance.addObserver(this);
+  }
+
+  /// Dispose the [WidgetsBinding] observer.
+  void dispose() {
+    WidgetsBinding.instance.removeObserver(this);
+  }
+
+  /// Plays and loops a background music file specified by [filename].
+  ///
+  /// The volume can be specified in the optional named parameter [volume]
+  /// where `0` means off and `1` means max.
+  ///
+  /// It is safe to call this function even when a current BGM track is
+  /// playing.
+  void play(String filename, {double volume}) async {
```

Comment:
why not just set the default value here? Like

```dart
  void play(String filename, {double volume = 1}) async {
```

Reply:
the first reason is that i didn't know i can do that :D
but i'm keeping it as is because of the second reason, which is: during development time, passing `null` to a `double` parameter is perfectly valid but could cause a run-time error. the null-aware assignment is a safer "catch code" making the default value in the parameter definition unnecessary.